### ISSUE TITLE
Unintended logic in util.behavorial_element_coordinates

### ISSUE THREAD

**Issue by basvdl97** [droefs/HLISA#40]
Looking at this part of `util.behavorial_element_coordinates`:

```python
for i in range(100): 
    x = x_relative + int(np.random.normal(int(element.rect['width']*0.5), int(element.rect['width']*0.2)))
    y = y_relative + int(np.random.normal(int(element.rect['height']*0.5), int(element.rect['height']*0.2)))
    coords_in_button = webdriver.execute_script(f"return document.elementFromPoint({x}, {y}) === arguments[0];", element)
    coords_in_descendant = webdriver.execute_script(f"""
        let el = document.elementFromPoint({x}, {y});
        return [...arguments[0].querySelectorAll('*')].includes(el);""", element)
    if x < 0 or y < 0 or x > viewport_width or y > viewport_height:
        coords_in_button = False # If the element is partly in the viewport, a part of the element is outside of it. In that case, try again. This is not the best solution (non-deterministic), it would be better to limit the sample space.
    if coords_in_button or coords_in_descendant:
        return (x, y)
```

This part selects possible random coordinates with in the element:

```python
x = x_relative + int(np.random.normal(int(element.rect['width']*0.5), int(element.rect['width']*0.2)))
y = y_relative + int(np.random.normal(int(element.rect['height']*0.5), int(element.rect['height']*0.2)))
```

Then we check if the coordinates are over the element or one of the element's descendants:
```python
coords_in_button = webdriver.execute_script(f"return document.elementFromPoint({x}, {y}) === arguments[0];", element)
coords_in_descendant = webdriver.execute_script(f"""
        let el = document.elementFromPoint({x}, {y});
        return [...arguments[0].querySelectorAll('*')].includes(el);""", element)
```

Here comes the part where I believe we intend to check if the `x, y` place we want to move to is not outside the viewport. And in order 'mark' that we are outside the viewport we set `coord_in_button` to false.

```python
if x < 0 or y < 0 or x > viewport_width or y > viewport_height:
    coords_in_button = False
```

However when we decide if we want to return the coordinates we check with an `or`:
```python
if coords_in_button or coords_in_descendant:
    return (x, y)
```

This results in that if the chosen `x, y` is a descendant element we don't care if it is outside of the viewport, and return the coordinates any ware.

I believe the intent of the viewport check is to make sure the `x, y` are in the viewport regardless of if the `x, y` is over the element or its descendants.

**Comment by basvdl97** [comment]
I propose the following fix that checks if the `x, y` are in the viewport right after generating them. And the Booleans `coords_in_button` and `coords_in_descendant` remain in use to check if the `x, y` are actually over the element.

```python
for _ in range(100): 
    x = x_relative + int(np.random.normal(int(element.rect['width']*0.5), int(element.rect['width']*0.2)))
    y = y_relative + int(np.random.normal(int(element.rect['height']*0.5), int(element.rect['height']*0.2)))
    
    if x < 0 or y < 0 or x > viewport_width or y > viewport_height:
        continue
    
    coords_in_button = webdriver.execute_script(f"return document.elementFromPoint({x}, {y}) === arguments[0];", element)
    coords_in_descendant = webdriver.execute_script(f"""
        let el = document.elementFromPoint({x}, {y});
        return [...arguments[0].querySelectorAll('*')].includes(el);""", element)

    if coords_in_button or coords_in_descendant:
        return (x, y)
```

**Comment by bkrumnow** [comment]
Hi, 
Thanks for your report. I think your fix should be merged. However, we should also refactor this and other parts in the util file. With more code clarity, we would have avoided making such mistakes.

**Comment by droefs** [comment]
Thank you for this contribution, your fix has been merged.

I agree that the util file can use some refactoring. Except the function `get_cursor_coordinates()` as it has changed already in my work-in-progress fix for #31.
